File: engines/v8/cpp/source/console.cpp
#include "klyron_v8.h"
#include "cpp/impl/internal.h"

#include <cstdio>
#include <cstring>
#include <ctime>
#include <string>

static std::string timestamp() {
    time_t now = time(nullptr);
    struct tm result;
    struct tm* t = localtime_r(&now, &result);
    char buf[64];
    strftime(buf, sizeof(buf), "%H:%M:%S", t);
    return std::string("[") + buf + "]";
}

static void console_output(const char* level, const char* msg, FILE* fp) {
    fprintf(fp, "%s %s: %s\n", timestamp().c_str(), level, msg);
    fflush(fp);
}

klyron_v8_value_t* klyron_v8_console_new(klyron_v8_context_t* ctx) {
    if (!ctx) return nullptr;
    auto iso = get_iso(ctx);
    v8::Locker locker(iso);
    v8::Isolate::Scope iso_scope(iso);
    v8::HandleScope scope(iso);
    auto context = get_ctx(ctx);
    v8::Context::Scope ctx_scope(context);

    auto obj = v8::Object::New(iso);

    auto log_fn = v8::Function::New(context,
        [](const v8::FunctionCallbackInfo<v8::Value>& info) {
            auto iso = info.GetIsolate();
            v8::HandleScope scope(iso);
            v8::String::Utf8Value str(iso, info[0]);
            if (*str) console_output("LOG", *str, stdout);
        }).ToLocalChecked();
    obj->Set(context, v8::String::NewFromUtf8(iso, "log", v8::NewStringType::kNormal).ToLocalChecked(), log_fn).Check();

    auto warn_fn = v8::Function::New(context,
        [](const v8::FunctionCallbackInfo<v8::Value>& info) {
            auto iso = info.GetIsolate();
            v8::HandleScope scope(iso);
            v8::String::Utf8Value str(iso, info[0]);
            if (*str) console_output("WARN", *str, stderr);
        }).ToLocalChecked();
    obj->Set(context, v8::String::NewFromUtf8(iso, "warn", v8::NewStringType::kNormal).ToLocalChecked(), warn_fn).Check();

    auto error_fn = v8::Function::New(context,
        [](const v8::FunctionCallbackInfo<v8::Value>& info) {
            auto iso = info.GetIsolate();
            v8::HandleScope scope(iso);
            v8::String::Utf8Value str(iso, info[0]);
            if (*str) console_output("ERROR", *str, stderr);
        }).ToLocalChecked();
    obj->Set(context, v8::String::NewFromUtf8(iso, "error", v8::NewStringType::kNormal).ToLocalChecked(), error_fn).Check();

    auto info_fn = v8::Function::New(context,
        [](const v8::FunctionCallbackInfo<v8::Value>& info) {
            auto iso = info.GetIsolate();
            v8::HandleScope scope(iso);
            v8::String::Utf8Value str(iso, info[0]);
            if (*str) console_output("INFO", *str, stdout);
        }).ToLocalChecked();
    obj->Set(context, v8::String::NewFromUtf8(iso, "info", v8::NewStringType::kNormal).ToLocalChecked(), info_fn).Check();

    auto debug_fn = v8::Function::New(context,
        [](const v8::FunctionCallbackInfo<v8::Value>& info) {
            auto iso = info.GetIsolate();
            v8::HandleScope scope(iso);
            v8::String::Utf8Value str(iso, info[0]);
            if (*str) console_output("DEBUG", *str, stdout);
        }).ToLocalChecked();
    obj->Set(context, v8::String::NewFromUtf8(iso, "debug", v8::NewStringType::kNormal).ToLocalChecked(), debug_fn).Check();

    return new klyron_v8_value(iso, obj, ctx->parent);
}

void klyron_v8_console_log(klyron_v8_context_t* ctx, const char* msg) {
    if (msg) console_output("LOG", msg, stdout);
}

void klyron_v8_console_warn(klyron_v8_context_t* ctx, const char* msg) {
    if (msg) console_output("WARN", msg, stderr);
}

void klyron_v8_console_error(klyron_v8_context_t* ctx, const char* msg) {
    if (msg) console_output("ERROR", msg, stderr);
}

void klyron_v8_console_info(klyron_v8_context_t* ctx, const char* msg) {
    if (msg) console_output("INFO", msg, stdout);
}

void klyron_v8_console_debug(klyron_v8_context_t* ctx, const char* msg) {
    if (msg) console_output("DEBUG", msg, stdout);
}

File: engines/v8/src/console.rs

#[cfg(feature = "native")]
use crate::ffi;
#[cfg(feature = "native")]
use std::ffi::CString;

pub struct V8Console {
    #[cfg(feature = "native")]
    ctx: *mut ffi::V8ContextHandle,
    #[cfg(feature = "native")]
    handle: Option<*mut ffi::V8ValueHandle>,
}

impl V8Console {
    #[cfg(feature = "native")]
    pub fn new(ctx: *mut ffi::V8ContextHandle) -> Self {
        let handle = unsafe { ffi::klyron_v8_console_new(ctx) };
        Self { ctx, handle: Some(handle) }
    }

    #[cfg(not(feature = "native"))]
    pub fn new(_ctx: *mut std::ffi::c_void) -> Self {
        Self {}
    }

    #[cfg(feature = "native")]
    pub fn log(&self, msg: &str) {
        if let Ok(c) = CString::new(msg) {
            unsafe { ffi::klyron_v8_console_log(self.ctx, c.as_ptr()) }
        }
    }

    #[cfg(feature = "native")]
    pub fn warn(&self, msg: &str) {
        if let Ok(c) = CString::new(msg) {
            unsafe { ffi::klyron_v8_console_warn(self.ctx, c.as_ptr()) }
        }
    }

    #[cfg(feature = "native")]
    pub fn error(&self, msg: &str) {
        if let Ok(c) = CString::new(msg) {
            unsafe { ffi::klyron_v8_console_error(self.ctx, c.as_ptr()) }
        }
    }

    #[cfg(feature = "native")]
    pub fn info(&self, msg: &str) {
        if let Ok(c) = CString::new(msg) {
            unsafe { ffi::klyron_v8_console_info(self.ctx, c.as_ptr()) }
        }
    }

    #[cfg(feature = "native")]
    pub fn debug(&self, msg: &str) {
        if let Ok(c) = CString::new(msg) {
            unsafe { ffi::klyron_v8_console_debug(self.ctx, c.as_ptr()) }
        }
    }

    #[cfg(feature = "native")]
    pub fn handle(&self) -> Option<*mut ffi::V8ValueHandle> {
        self.handle
    }
}
